### DISCUSSION TITLE
Open Source Craft

### DISCUSSION THREAD

**Discussion by angrybrad** [craftcms/cms#8114] [Ideas]
> Created by: Nathan Doyle (natetronn@gmail.com) on 2015/03/11 04:20:14 +0000
> Votes at time of UserVoice import: 44

-----------------------------------

The title says it all!

This idea for this came from Brandon in fact (to a degree anyway) when we were talking about plugin pricing and business models. He stated, and I'm paraphrasing, if revenue from the Plugin Store commissions and Craft Commerce (presumably) reached a point of success it would allow P&T to lower the price of Craft itself.

I'd like to see this happen and hope such business models could be implemented with such success that Craft could become FOSS FTW!!! 

(yes I triple exclamation pointed)

Here is what BK said: http://monosnap.com/image/OKgkwHqfN8EXu8ALcIzPlbvkKOX9sk

**Comment by angrybrad** [comment]
> Posted by Clive Portman (clive@cliveportman.co.uk) on 2015/10/13 07:13:43 +0000
-----------------------------------

Part of the reason I use Craft is because it isn&#39;t OSS - I like the fact someone is accountable for looking after the platform. I want Pixel &amp; Tonic to focus on developing the CMS as their core business, not client work and definitely not supporting their business through selling plugins (beyond a couple of irreplaceable ones like Craft Commerce).

One of the biggest problems with Wordpress is the number of plugins required, with the associated security issues. One of Craft&#39;s strengths is that so much functionality is provided in the core software, the idea being we need fewer plugins, reducing our maintenance costs and improving the security of our sites. This is a big sell for us when clients are already familiar with Wordpress, for example. Where possible, I&#39;d like to see popular plugins make it into core.

If the business model switches to plugin revenue, I&#39;d be very concerned about that. Sorry.

**Comment by angrybrad** [comment]
> Posted by PlutoMars (Limey.pluto@gmail.com) on 2015/03/20 09:37:19 +0000
-----------------------------------

I&#39;ve been noticing that pubic bodies, tendering bigger webprojects require often &quot;an open source CMS&quot;, and they often then end-up using Drupal. So in this regard, open-sourcing it would be definitely a big plus. As it looks like we&#39;re going to have to learn Drupal now...

**Comment by angrybrad** [comment]
> Posted by Brandon Kelly (brandon@pixelandtonic.com) on 2015/03/06 03:54:45 +0000
-----------------------------------

@Nate That&#39;s actually very close to something we are considering: the Open Core approach where Craft&#39;s underlying Element architecture, plugin APIs, and user accounts all get released as a self-contained OSS platform (Craft Core, Craft CMF, or something like that), and that just becomes one more Composer dependency for the still-proprietary Craft CMS, which would continue to be released the same way it is today.

The nice thing about this approach is, it will have little (if any) effect on Craft sales so we don&#39;t have to worry about cash flow, but should put many developers at ease about developing on top of a proprietary platform, as a much larger percentage of it (and basically all of the code plugins are touching) will be OSS.

**Comment by angrybrad** [comment]
> Posted by Nate Iler (nate@flipboxdigital.com) on 2015/03/11 14:29:14 +0000
-----------------------------------

Interesting topic...

I believe a few days ago there was a short thread running in the Craft Slack #craft3 channel around re-writing Craft into isolated modules (Users, Categories, Tags, etc) which you could elect to install via a config/composer file which I found exciting.  Great concept.  It’s not exactly open source, but it’s opening the codebase.

I’m so used to an entity controlling the code that the appeal of making it OOS doesn’t weigh in on me.  There is a forum, such as this, in which the community can voice an opinion...and yes there are a few aspects of Craft that I would have love to perform a pull request against, but don’t lose sleep over it.

There are applications such as Magento which have an open source and commercial approach which is interesting.  An enterprise license there will run your client 10k (and to sell a license you have to pay Magento to become a partner). Magento Community will give you the building blocks, but enhanced caching, segmentation, etc are all in the Enterprise edition.

Craft was created from a concept and vision of P&amp;T, which I think many of us are aligned with.  After all, Craft is built on Yii, an open source framework, and according to the upcoming 3.0 release will allow for plugin development at the module level that Craft itself is writing on.  In theory, the Craft community can leverage a foundation to ‘craft’ functionality they desire.  When I consider Craft, I’m also evaluating the stable vision behind it…is it aligned with mine.  A caveat, however, is whether P&amp;T is listening to the community…and my evaluation is yes.

Would the community entertain P&amp;T taking the core Craft ‘app’ + ‘Element’ and making it FOSS while converting each of their (current) first party Elements (Users, Categories, Tags, etc) to the plugin store where you can purchase them at $100/each?  I think I would.

**Comment by angrybrad** [comment]
> Posted by David Ritter (david.ritter@spinthis.com) on 2015/05/18 19:35:25 +0000
-----------------------------------

Some off-the-cuff thoughts here....


Let’s not forget that Craft has ultimately succeeded in a market where many aspects of web design have really become commoditized, and many options are low cost or even free. As long as Craft is making the company money, it would be stupid to throw that away. It would be a different story if you hardly moved any Craft licenses. 

I think Craft still has a bit of growing to do. It would be a mistake to open-source it at-the-moment but that&#39;s just my general sentiment.  I don’t think there’s money in open source unless you got something else going on. Envato has shown that plenty of people are willing to pay for software, CodeCanyon and ThemeForest are 2 great examples. 

Let&#39;s not forget that Rails came out of Basecamp and CodeIgnitor came from EE. WebKit came from Apple.    All of those projects had a product to sell on top of what they “gave away”. Apple had Macs (and down the road iPhones) to sell, for example. Most people using BaseCamp wouldn’t be building a competitor to it. Yet in the web market, I think everyone knows someone who have tried to build their own CMS or framework—if they haven’t done so themselves already. 

If Craft were open source, where’s the motivation—read revenue—to keep improving the code base? P&amp;T’s resources would inevitably shift to making sure it can sustain itself.  OSS tends to work well when the project is pretty well established and has some company backing as a mutual benefactor, sometimes selling a service that doesn’t directly rely on its OSS offering.   That usually revolves around ancillary services like add-ons and providing developer resources like documentation,  support and probably doing something like a SaaS offering. I don’t know if something like Craft commerce and others would be enough.  Instead of writing Craft, P&amp;T would be writing Craft add-ons. That’s just the nature of the beast.

How many OSS projects go into ruin because they’re just not funded? It happens all the time. The big issue I see is making sure all of P&amp;T’s hard work on this project doesn’t get squandered if it were to go OSS. Once you go that way,  it’s hard to go the other way.

Apple used to sell clones back in the early ‘90s. Apple thought that by licensing Mac OS to other hardware manufacturers, that  would increase its market share of Mac OS and thus income. Instead it robbed Apple of income it relied on by selling its own hardware.  Apple more or less forgot what market they were in. Make no mistake, Apple is in the hardware business, regardless of whatever awesome software they make. (For what it’s worth Microsoft is now having the same problem. They can’t give the software away—Windows 10 will be a free upgrade BTW.)

What business is P&amp;T in? I see P&amp;T being in the CMS business. They are a product design company. If it shifted over to more of a support role, I just can’t see that working. Brandon and company have a real niche for solving customer problems. It’d be like Apple giving away its iPhones away for free and relying on App Store developers for income. There’s too many variables and uncertainly there to make a business out of someone else’s scraps.

Developers tend to be a very vocal minority and some of them very bad business people and sales people.  I doubt open-sourcing Craft will really help or hurt the adoption or not, regardless of the feedback. If you’re a developer and complaining about cost, I can’t see you lasting for long in this industry.  Brandon nailed it when he said “[agencies are] building custom applications for their clients to serve complex needs.”   

The agencies spitting out $20k+ custom WordPress websites obviously don’t care about cost so I think that it’s really a red herring argument. $300 is what, 3 billable hours? What I think these companies/developers are scared of is trusting a small, relatively unproven company no matter what the cost.  Open source is basically a scapegoat. “It’s free so there’s no risk!”  It an easy argument to win. I would bet dollars to donuts those same developers complaining about Craft not being &quot;open&quot; are probably having problems financially, not charging enough for their work, etc.

For my company at least, I’m happy to help put food on the table for a company that helps put food on mine. Open source of not, I want my clients to have the best possible experience and Craft helps me do that.  I think open source has its place. I love Ubuntu and the LAMP stack has ultimately been a very positive thing as a whole. They just work. But Craft isn&#39;t there yet.

**Comment by angrybrad** [comment]
> Posted by Nathan Doyle (natetronn@gmail.com) on 2015/03/11 04:20:14 +0000
-----------------------------------

One other thought about the EE add-ons sales. P&amp;T may in fact have insights into sales as it relates to EE and could possibly infer allot of things from that community. In fact, what happened (or didn&#39;t happen) over at EE/EL is exactly why I&#39;m super excited to see this even being talked about.

A few years ago ExpressionEngine had the technical advantage against the big three but, since they weren&#39;t OSS among other things, watched as their market share declined and the big three&#39;s market share continued to rise at staggering rates. Other more niche and current tech CMS continued to be released at the same time.

Small sites which everyone use to build on EE were now being built on Statamic. Instead of fighting the fact that EE wasn&#39;t OSS sites started moving over to Processwire. People who were so fed up with decision over there jumped ship completely and landed, full speed ahead I might add, over here with Craft and P&amp;T. And that&#39;s *Straight Up* the truth!

Craft is coming in at very similar place if you ask me. It&#39;s a superior core product form a technical point of view but, I fear, will end in a similar fate that EE did unless P&amp;T is not careful. I&#39;m not trying to &quot;scare&quot; anyone into OSS and I realize this is all just hearsay, as I can&#39;t predict the future but, it&#39;s stuff that I think about so, thought I&#39;d share it just the same.

**Comment by angrybrad** [comment]
> Posted by Nathan Doyle (natetronn@gmail.com) on 2015/03/11 04:20:14 +0000
-----------------------------------

Someone mentioned licensing keeps out the &quot;riff raff&quot; and by open sourcing Craft it would open up the flood gates so to speak. I can see this causing some issues in that regard but, we all started someone where yes? Where would you prefer people to start? Here or over at [fill in the black]?

By having a clear~ trajectory of where the community is heading, by steering it in a positive direction etc. shouldn&#39;t it be possible to pull people up along with us instead of pushing them down and away?

The code base could keep allot of issues to a minimum either way in that regard. It&#39;s not procedural for one, the learning curve is quite a bit steeper and Craft is vanilla so requires a depth of knowledge that extends to a greater range of skill sets besides just Craft knowhow itself. Plus, as others have suggested, having a business/support model in place can hopefully alleviate this for P&amp;T if done right.

For me the argument about free or not isn&#39;t the actual &quot;problem.&quot; Not that it&#39;s free or not. I could personally care less about Craft&#39;s current price tag. Clients buy their own licenses. I don&#39;t even own one at this point. So free for me it isn&#39;t about the price of the product or OSS for that matter rather, from a purely competitive point of view, it could appear to level the playing ground. This is psychology we are talking about not if something is really free or not. We all know the 4 hours worth of time (or whatever) we gain back for the client in the long run by using Craft completely offsets the cost of the license (starting from zero that is. Theme re-skinning is a different topic of course.)

The more things being equal, from a strictly non-technical point of view, the more Craft as a framework or platform i.e. the technical advantages can really stand out and shine yes?

So again, my motives, as a client (technically I&#39;m not) has nothing to do with money and more to do with psychology.

(I&#39;d love to hear the reverse of this type of thinking. Totally open to have my thought processes questioned so I can reevaluate them.)

By the way, I recommend Craft to people over on Twitter all the time. I&#39;m sure P&amp;T is fully aware of this fact. Every other time I do someone mentions that Craft looks GREAT but, and the but is usually the fact that it&#39;s not OSS. At that point I usually go down a list of other similar CMS etc. that are OSS. For some it&#39;s about the money and Craft not being free, yes I agree but, most of the people I talk with have clients who can afford to pay the license fee 20 times over so, it&#39;s about something completely else I can assure you.

With that said, I am in agreement about what OP said in regards to not taking everything your clients say and run with it etc. As I mentioned in the start, I&#39;m super excited to see this conversation even up for debate. And by debate I mean talked about to the extent that it&#39;s being talked about. 

This whole thing might die here and not be mentioned again for another year or two; when they are in a position to even think about such a thing. It might die here and never come up again and that&#39;s cool too. Until P&amp;T comes out and officially says &quot;We are looking into open sourcing Craft and need feedback&quot; or something, then this whole topic should be considered us just having a discussion for the fun of it.

**Comment by angrybrad** [comment]
> Posted by Luke Stevens (luke@lukestevensdesign.com) on 2015/05/23 01:33:28 +0000
-----------------------------------

@Brandon, no worries :) Few more thoughts...

For sure, I think we&#39;d all love to see Craft (along with P&amp;T) grow substantially, and undoubtedly there&#39;s a bunch of positives that come with that like you say - more resources for you guys to throw at the product; more devs building on it; more people releasing add-ons; more support &amp; educational resources; more Craft-related work for web shops; a bigger, stronger community, etc. And if open source is the ticket there, awesome :) I just want to encourage the community to remember that what&#39;s in their (perceived) interests (free stuff yay!) is not *by definition* in P&amp;T&#39;s interests (though it might be), hence the theoretical other market opportunities. But if &quot;It&#39;s not open source!&quot; is a recurring objection, then you&#39;d be the ones hearing it, for sure :)

(On the developer adoption side, that&#39;s very interesting. There&#39;s some *very* interesting value prop ideas within your observation that &quot;The agencies that have been the most successful with Craft are the ones that are delivering more than just &quot;static&quot; websites; they&#39;re building custom applications for their clients to serve complex needs.&quot; btw. Very interesting to think about what customers &#39;hire&#39; Craft for, to borrow the JTBD lingo.)

If open source is a significant obstacle, then I guess it&#39;s worth considering:

- Forks &amp; governance: This seems like an issue that plagues just about every OS project (e.g. CI), so licensing rules around distribution and governance would have to be very carefully thought through. (The recent node.js vs io.js case being a good example.) Rails obviously worked well with a benevolent dictator, and I&#39;m sure you&#39;d do well there. But drama + tears seem to inevitably arise around these things at some point, so I&#39;d be inclined to tread carefully :) 
- Freemium or true open source? Craft is already kind-of freemium I suppose, but to go fully freemium/open source obviously entails some sort of paid/premium offering, and I imagine you could test what paid support (for e.g.) looks like for companies where the license fee of Craft is negligible relative to the size of their business, and if that takes off, then consider dropping the Craft licensing fee down the track. In this case, going open source is a nice potential side benefit of an alternative business model, but it&#39;s a side effect of success, not an (initial) driver of that success. (Of course, if open source does drive the premium support flywheel down the track, then great!) Here the question is not &quot;Should we go open source?&quot; (irrelevant for now) but &quot;Can we build a business-supporting recurring revenue stream outside of Craft licenses?&quot; and that&#39;s a pretty interesting question :)
- You could also look at Highchart&#39;s business model if you haven&#39;t already, it seems to have worked pretty well for them: http://shop.highsoft.com/highcharts.html . Free to d/l &amp; try (free for non-profits I think), small fee for one-off licenses, higher fees for application development, and then support packages with very clear support offerings (e.g. number of hours &amp; availability etc). That captures more value in licensing costs for higher-end projects, and has very clear costs around support. Naturally, the product is excellent &amp; developers sing its praises too :) And it&#39;s just one of many models too, of course.
- I&#39;m not sure how good your P&amp;T EE add-on sales are as a proxy of EE&#39;s sales in general, but if (if!) they are somewhat indicative, it may be interesting to look back at your sales vs EE&#39;s business model changes and see if there were any interest correlations there, if you haven&#39;t already (though they obviously kept their license price). Tricky to tease out causation for sure, but it might provide another small data point as you weigh up your options. 

Anyway, just a few reflective questions &amp; ideas to hopefully help you think through the issue :)

**Comment by angrybrad** [comment]
> Posted by Kevin (kevin@pixelcamp.co.uk) on 2015/05/23 18:48:08 +0000
-----------------------------------

I (personally) am comforted by the fact that a &#39;body&#39; is accountable for a product and also has an incentive to offer great support and development via increasing sales (usually well-earnt), nothing against open-source model at all, but with Craft being a steal anyway and the advantages of the above - if it ain&#39;t broke et al…

**Comment by angrybrad** [comment]
> Posted by Gregory Stewart (gstewart512@gmail.com) on 2015/05/23 15:06:11 +0000
-----------------------------------

My company is spending significant resources on relaunching our web presence on Craft, and we would be very interested in some sort of VIP support plan as mentioned in the thread. I am sure other companies would be as well.

**Comment by frob** [comment]
With all the comments being posted here referencing some sort of comfort in the fact that a body is responsible for the source code. That is already the purview of the service agreement and not the source code license. 

From the CreaftCMS [source code license](https://github.com/craftcms/cms/blob/develop/LICENSE.md).

> THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES, OR OTHER LIABILITY, INCLUDING SPECIAL, INCIDENTAL AND CONSEQUENTIAL DAMAGES, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

**Comment by brandonkelly** [comment]
@frob Right, if Craft source code were to go MIT, we would absolutely still own its branding and the master repository, and maintain our status as its chief maintainers.

**Comment by frob** [comment]
If Craft went GPL then any derivative would also be forced to be open source. This would further protect the source project (Craft) and allow for derivatives to be rolled back into Craft. MIT has its benefits too, but I am no lawyer and you will definitely want to talk to one of those before re-licensing your project.

I will have to say, as someone who uses OSS for most of my work, it is always difficult when I am forced to use something that is closed source. I am always coming across stuff where I think "I could fix this. No wait, its closed source."

**Comment by brandonkelly** [comment]
GPL would require all plugins and many websites built with Craft to also be GPL, and released publicly. So that’s a non-starter :)

> I am always coming across stuff where I think "I could fix this. No wait, its closed source."

Craft isn’t closed _source_ – all the source code is right here on GitHub, and the [license](https://github.com/craftcms/cms/blob/develop/LICENSE.md) is pretty unrestrictive as proprietary software goes. Want to change something? Fork it. Want to push the change back up to us? Submit a PR. Just like any other (F)OSS project.

**Comment by frob** [comment]
> GPL would require all plugins and many websites built with Craft to also be GPL, and released publicly. So that’s a non-starter.

This is an incorrect assumption. If this where true, no site could be built on WordPress or Drupal. Currently whitehouse.gov is built with WordPress, but you cannot download the source code. GPL code can be used with non-GLP code and any entity can use GPL code (so long as they are not redistributing it) without releasing the source code. Plugins don't have to be GLPed either, though the marketplace economy might complicate things.

> Craft isn’t closed source – all the source code is right here on GitHub, and the license is pretty unrestrictive as proprietary software goes. Want to change something? Fork it. Want to push the change back up to us? Submit a PR. Just like any other (F)OSS project.

Yes, don't get me wrong. That statement didn't refer you this project. Trust me that I would be submitting PRs if I was using CraftCMS. I am a little hesitant to recommend Craft with a custom open source license. Even if this was a MIT+ license that added the license and payment, much like React used to have. I am hesitant because the standard open source licenses have been tested. 

Please don't take this a harsh criticism. From what I have seen, CraftCMS is well done and I was stoked to see the source available for download and PRs being accepted.

**Comment by ryanpcmcquen** [comment]
> GPL would require all plugins and many websites built with Craft to also be GPL, and released publicly. So that’s a non-starter :)

@brandonkelly, have you considered the MPL license?

**Comment by brandonkelly** [comment]
@ryanpcmcquen No but that’s an interesting one!

**Comment by glenux** [comment]
Ok, from what I understand so far, CraftCMS is not (yet) open source, so who has the intellectual property for the resulting mix of all that contributed code & PRs ?

**Comment by brandonkelly** [comment]
It’s commercial open source, just not FOSS (yet anyway). All contributed IP is owned by us, per the [contributing notes](https://github.com/craftcms/cms/blob/develop/CONTRIBUTING.md).

**Comment by mgifford** [comment]
@brandonkelly what do you gain by owning all of the contributed IP? How does that fit into your business model? 

I'm a fan of the GPL, but if you are releasing the code and wanting to scale the project with more users and developers, it just seems that owning the IP is just a default condition that you've fallen into.  

Maybe I'm wrong. You'd absolutely want to hold onto the branding and manage control of the repo.  None of that is in question as far as I'm concerned.

**Comment by brandonkelly** [comment]
@mgifford The right to do what we want with the code w/out needing to get permission from each contributor. That’s the way all proprietary-licensed software works.

If we are ever able go FOSS with Craft ([working on it](https://twitter.com/brandonkelly/status/1279121533076746240)), that will no longer be the case.

**Comment by mustafa-online** [comment]
is that still a thing? FOSS? or with less price?

**Comment by vitalijalbu** [comment]
guys, at least provide user auth inside the box, not with pricing plans.
